File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/EmitterConfiguration.java
package com.snowplowanalytics.snowplow.configuration;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.emitter.BufferOption;
import com.snowplowanalytics.snowplow.internal.emitter.EmitterConfigurationInterface;
import com.snowplowanalytics.snowplow.network.RequestCallback;
import com.snowplowanalytics.snowplow.emitter.EventStore;

import java.util.Map;

/**
 * It allows the tracker configuration from the emission perspective.
 * The EmitterConfiguration can be used to setup details about how the tracker should treat the events
 * to emit to the collector.
 */
public class EmitterConfiguration implements Configuration, EmitterConfigurationInterface {

    /**
     * @see #bufferOption(BufferOption)
     */
    @NonNull
    public BufferOption bufferOption;

    /**
     * @see #emitRange(int)
     */
    public int emitRange;

    /**
     * @see #threadPoolSize(int) 
     */
    public int threadPoolSize;

    /**
     * @see #byteLimitGet(int) 
     */
    public long byteLimitGet;

    /**
     * @see #byteLimitPost(int) 
     */
    public long byteLimitPost;

    /**
     * @see #requestCallback(RequestCallback) 
     */
    @Nullable
    public RequestCallback requestCallback;

    /**
     * @see #eventStore(EventStore)
     */
    @Nullable
    public EventStore eventStore;

    /**
     * @see #customRetryForStatusCodes(Map)
     */
    @Nullable
    public Map<Integer, Boolean> customRetryForStatusCodes;

    /**
     * @see #serverAnonymisation(boolean)
     */
    public boolean serverAnonymisation;

    // Constructor

    /**
     * It sets a default EmitterConfiguration.
     * Default values:
     *         bufferOption = BufferOption.Single;
     *         emitRange = 150;
     *         threadPoolSize = 15;
     *         byteLimitGet = 40000;
     *         byteLimitPost = 40000;
     *         serverAnonymisation = false;
     */
    public EmitterConfiguration() {
        bufferOption = BufferOption.Single;
        emitRange = 150;
        threadPoolSize = 15;
        byteLimitGet = 40000;
        byteLimitPost = 40000;
        serverAnonymisation = false;
    }

    // Getters and Setters

    @Override
    @Nullable
    public EventStore getEventStore() {
        return eventStore;
    }

    @Override
    @NonNull
    public BufferOption getBufferOption() {
        return bufferOption;
    }

    @Override
    public void setBufferOption(@NonNull BufferOption bufferOption) {
        this.bufferOption = bufferOption;
    }

    @Override
    public int getEmitRange() {
        return emitRange;
    }

    @Override
    public void setEmitRange(int emitRange) {
        this.emitRange = emitRange;
    }

    @Override
    public int getThreadPoolSize() {
        return threadPoolSize;
    }

    public void setThreadPoolSize(int threadPoolSize) {
        this.threadPoolSize = threadPoolSize;
    }

    @Override
    public long getByteLimitGet() {
        return byteLimitGet;
    }

    @Override
    public void setByteLimitGet(long byteLimitGet) {
        this.byteLimitGet = byteLimitGet;
    }

    @Override
    public long getByteLimitPost() {
        return byteLimitPost;
    }

    @Override
    public void setByteLimitPost(long byteLimitPost) {
        this.byteLimitPost = byteLimitPost;
    }

    @Override
    @Nullable
    public RequestCallback getRequestCallback() {
        return requestCallback;
    }

    @Override
    public void setRequestCallback(@Nullable RequestCallback requestCallback) {
        this.requestCallback = requestCallback;
    }

    @Nullable
    @Override
    public Map<Integer, Boolean> getCustomRetryForStatusCodes() {
        return customRetryForStatusCodes;
    }

    @Override
    public void setCustomRetryForStatusCodes(@Nullable Map<Integer, Boolean> customRetryForStatusCodes) {
        this.customRetryForStatusCodes = customRetryForStatusCodes;
    }

    @Override
    public boolean isServerAnonymisation() {
        return serverAnonymisation;
    }

    @Override
    public void setServerAnonymisation(boolean serverAnonymisation) {
        this.serverAnonymisation = serverAnonymisation;
    }

    // Builders

    /**
     * Sets whether the buffer should send events instantly or after the buffer
     * has reached it's limit. By default, this is set to BufferOption Default.
     */
    @NonNull
    public EmitterConfiguration bufferOption(@NonNull BufferOption bufferOption) {
        this.bufferOption = bufferOption;
        return this;
    }

    /**
     * Maximum number of events collected from the EventStore to be sent in a request.
     */
    @NonNull
    public EmitterConfiguration emitRange(int emitRange) {
        this.emitRange = emitRange;
        return this;
    }

    /**
     * Maximum number of threads working in parallel in the tracker to send requests.
     */
    @NonNull
    public EmitterConfiguration threadPoolSize(int threadPoolSize) {
        this.threadPoolSize = threadPoolSize;
        return this;
    }

    /**
     * Maximum amount of bytes allowed to be sent in a payload in a GET request.
     */
    @NonNull
    public EmitterConfiguration byteLimitGet(int byteLimitGet) {
        this.byteLimitGet = byteLimitGet;
        return this;
    }

    /**
     * Maximum amount of bytes allowed to be sent in a payload in a POST request.
     */
    @NonNull
    public EmitterConfiguration byteLimitPost(int byteLimitPost) {
        this.byteLimitPost = byteLimitPost;
        return this;
    }

    /**
     * Custom component with full ownership for persisting events before to be sent to the collector.
     * If it's not set the tracker will use a SQLite database as default EventStore.
     */
    @NonNull
    public EmitterConfiguration eventStore(@Nullable EventStore eventStore) {
        this.eventStore = eventStore;
        return this;
    }

    /**
     * Callback called for each request performed by the tracker to the collector.
     */
    @NonNull
    public EmitterConfiguration requestCallback(@Nullable RequestCallback requestCallback) {
        this.requestCallback = requestCallback;
        return this;
    }

    /**
     * Custom retry rules for HTTP status codes returned from the Collector.
     * The dictionary is a mapping of integers (status codes) to booleans (true for retry and false for not retry).
     */
    @NonNull
    public EmitterConfiguration customRetryForStatusCodes(@Nullable Map<Integer, Boolean> customRetryForStatusCodes) {
        this.customRetryForStatusCodes = customRetryForStatusCodes;
        return this;
    }

    /**
     * Whether to anonymise server-side user identifiers including the `network_userid` and `user_ipaddress`
     */
    @NonNull
    public EmitterConfiguration serverAnonymisation(boolean serverAnonymisation) {
        this.serverAnonymisation = serverAnonymisation;
        return this;
    }

    // Copyable

    @Override
    @NonNull
    public EmitterConfiguration copy() {
        EmitterConfiguration copy = new EmitterConfiguration();
        copy.bufferOption = bufferOption;
        copy.emitRange = emitRange;
        copy.threadPoolSize = threadPoolSize;
        copy.byteLimitGet = byteLimitGet;
        copy.byteLimitPost = byteLimitPost;
        copy.eventStore = eventStore;
        copy.requestCallback = requestCallback;
        copy.customRetryForStatusCodes = customRetryForStatusCodes;
        copy.serverAnonymisation = serverAnonymisation;
        return copy;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/EmitterConfiguration.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.configuration

import com.snowplowanalytics.core.emitter.EmitterConfigurationInterface
import com.snowplowanalytics.core.emitter.EmitterDefaults
import com.snowplowanalytics.snowplow.emitter.BufferOption
import com.snowplowanalytics.snowplow.emitter.EventStore
import com.snowplowanalytics.snowplow.network.RequestCallback

/**
 * Configure how the tracker should send the events to the collector.     
 * 
 * Default values:
 *   - bufferOption: [BufferOption.DefaultGroup]
 *   - serverAnonymisation: false
 *   - emitRange: 150 - maximum number of events to process at a time
 *   - threadPoolSize: 15
 *   - byteLimitGet: 40000 bytes
 *   - byteLimitPost: 40000 bytes
 */
open class EmitterConfiguration : Configuration, EmitterConfigurationInterface {

    private var _isPaused: Boolean? = null
    internal var isPaused: Boolean
        get() = _isPaused ?: sourceConfig?.isPaused ?: false
        set(value) { _isPaused = value }

    /**
     * Fallback configuration to read from in case requested values are not present in this configuration.
     */
    internal var sourceConfig: EmitterConfiguration? = null

    private var _bufferOption: BufferOption? = null
    override var bufferOption: BufferOption
        get() = _bufferOption ?: sourceConfig?.bufferOption ?: BufferOption.DefaultGroup
        set(value) { _bufferOption = value }

    private var _emitRange: Int? = null
    override var emitRange: Int
        get() = _emitRange ?: sourceConfig?.emitRange ?: EmitterDefaults.emitRange
        set(value) { _emitRange = value }

    private var _threadPoolSize: Int? = null
    override var threadPoolSize: Int
        get() = _threadPoolSize ?: sourceConfig?.threadPoolSize ?: EmitterDefaults.threadPoolSize
        set(value) { _threadPoolSize = value }

    private var _byteLimitGet: Long? = null
    override var byteLimitGet: Long
        get() = _byteLimitGet ?: sourceConfig?.byteLimitGet ?: EmitterDefaults.byteLimitGet
        set(value) { _byteLimitGet = value }

    private var _byteLimitPost: Long? = null
    override var byteLimitPost: Long
        get() = _byteLimitPost ?: sourceConfig?.byteLimitPost ?: EmitterDefaults.byteLimitPost
        set(value) { _byteLimitPost = value }

    private var _requestCallback: RequestCallback? = null
    override var requestCallback: RequestCallback?
        get() = _requestCallback ?: sourceConfig?.requestCallback
        set(value) { _requestCallback = value }

    private var _eventStore: EventStore? = null
    override var eventStore: EventStore?
        get() = _eventStore ?: sourceConfig?.eventStore
        set(value) { _eventStore = value }

    private var _customRetryForStatusCodes: Map<Int, Boolean>? = null
    override var customRetryForStatusCodes: Map<Int, Boolean>?
        get() = _customRetryForStatusCodes ?: sourceConfig?.customRetryForStatusCodes
        set(value) { _customRetryForStatusCodes = value }

    private var _serverAnonymisation: Boolean? = null
    override var serverAnonymisation: Boolean
        get() = _serverAnonymisation ?: sourceConfig?.serverAnonymisation ?: EmitterDefaults.serverAnonymisation
        set(value) { _serverAnonymisation = value }
    
    // Builders
    
    /**
     * How many events to send in each request. By default, this is set to [BufferOption.DefaultGroup], 
     * a maximum of 10 events per request.
     */
    fun bufferOption(bufferOption: BufferOption): EmitterConfiguration {
        this.bufferOption = bufferOption
        return this
    }

    /**
     * Maximum number of events collected from the EventStore to be processed into requests at one time. 
     * The number of events per request is set using [EmitterConfiguration.bufferOption].
     */
    fun emitRange(emitRange: Int): EmitterConfiguration {
        this.emitRange = emitRange
        return this
    }

    /**
     * Maximum number of threads working in parallel in the tracker to send requests.
     */
    fun threadPoolSize(threadPoolSize: Int): EmitterConfiguration {
        this.threadPoolSize = threadPoolSize
        return this
    }

    /**
     * Maximum amount of bytes allowed to be sent in a payload in a GET request.
     */
    fun byteLimitGet(byteLimitGet: Int): EmitterConfiguration {
        this.byteLimitGet = byteLimitGet.toLong()
        return this
    }

    /**
     * Maximum amount of bytes allowed to be sent in a payload in a POST request.
     */
    fun byteLimitPost(byteLimitPost: Int): EmitterConfiguration {
        this.byteLimitPost = byteLimitPost.toLong()
        return this
    }

    /**
     * Custom component with full ownership for persisting events before to be sent to the collector.
     * If it's not set the tracker will use a SQLite database as 
     * [default EventStore](com.snowplowanalytics.core.emitter.storage.SQLiteEventStore).
     */
    fun eventStore(eventStore: EventStore?): EmitterConfiguration {
        this.eventStore = eventStore
        return this
    }

    /**
     * Callback called for each request performed by the tracker to the event collector.
     */
    fun requestCallback(requestCallback: RequestCallback?): EmitterConfiguration {
        this.requestCallback = requestCallback
        return this
    }

    /**
     * Custom retry rules for HTTP status codes returned from the collector.
     * The dictionary is a mapping of integers (status codes) to booleans (true for retry and false 
     * for not retry). By default, events in requests that return codes 400, 401, 403, 410, or 422 are
     * not retried and are deleted from the EventStore.
     */
    fun customRetryForStatusCodes(customRetryForStatusCodes: Map<Int, Boolean>?): EmitterConfiguration {
        this.customRetryForStatusCodes = customRetryForStatusCodes
        return this
    }

    /**
     * Whether to anonymise server-side user identifiers including the `network_userid` and `user_ipaddress`
     */
    fun serverAnonymisation(serverAnonymisation: Boolean): EmitterConfiguration {
        this.serverAnonymisation = serverAnonymisation
        return this
    }

    // Copyable
    override fun copy(): EmitterConfiguration {
        return EmitterConfiguration()
            .bufferOption(bufferOption)
            .emitRange(emitRange)
            .threadPoolSize(threadPoolSize)
            .byteLimitGet(byteLimitGet.toInt())
            .byteLimitPost(byteLimitPost.toInt())
            .eventStore(eventStore)
            .requestCallback(requestCallback)
            .customRetryForStatusCodes(customRetryForStatusCodes)
            .serverAnonymisation(serverAnonymisation)
    }
}
